sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp

```diff
@@ -0,0 +1,140 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+// Extra run to check for leaks in Level Zero using UR_L0_LEAKS_DEBUG
+// RUN: %if level_zero %{env SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=0 %{l0_leak_check} %{run} %t.out 2>&1 | FileCheck %s --implicit-check-not=LEAK %}
+// Extra run to check for immediate-command-list in Level Zero
+// RUN: %if level_zero %{env SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=1 %{l0_leak_check} %{run} %t.out 2>&1 | FileCheck %s --implicit-check-not=LEAK %}
+
+// Checks that the transitive queue recording feature is working as expected.
+// i.e. submitting a command group function to a queue that has a dependency
+// from a graph, should change the state of the queue to recording mode.
+
+#include "../graph_common.hpp"
+
+int main() {
+  queue Q1;
+  queue Q2;
+  queue Q3;
+
+  using T = int;
+
+  std::vector<T> DataA(Size), DataB(Size), DataC(Size);
+
+  std::iota(DataA.begin(), DataA.end(), 1);
+  std::iota(DataB.begin(), DataB.end(), 10);
+  std::iota(DataC.begin(), DataC.end(), 1000);
+
+  std::vector<T> ReferenceA(DataA), ReferenceB(DataB), ReferenceC(DataC);
+
+  T *PtrA = malloc_device<T>(Size, Q1);
+  T *PtrB = malloc_device<T>(Size, Q1);
+  T *PtrC = malloc_device<T>(Size, Q1);
+
+  Q1.copy(DataA.data(), PtrA, Size);
+  Q1.copy(DataB.data(), PtrB, Size);
+  Q1.copy(DataC.data(), PtrC, Size);
+  Q1.wait_and_throw();
+
+  exp_ext::command_graph Graph{Q1.get_context(), Q1.get_device()};
+
+  Graph.begin_recording(Q1);
+
+  auto GraphEventA = Q1.submit([&](handler &CGH) {
+    CGH.parallel_for(range<1>(Size), [=](item<1> Id) { PtrA[Id]++; });
+  });
+
+  // Since there is a dependency on GraphEventA which is part of a graph,
+  // this will change Q2 to the recording state.
+  auto GraphEventB = Q2.submit([&](handler &CGH) {
+    CGH.depends_on(GraphEventA);
+    CGH.parallel_for(range<1>(Size), [=](item<1> Id) { PtrA[Id]++; });
+  });
+
+  // Has no dependencies but should still be recorded to the graph because
+  // the queue was implicitly changed to recording mode previously.
+  auto GraphEventC = Q2.submit([&](handler &CGH) {
+    CGH.parallel_for(range<1>(Size), [=](item<1> Id) { PtrB[Id]++; });
+  });
+
+  // Q2 is now in recording mode. Submitting a command group to Q3 with a
+  // dependency on an event from Q2 should change it to recording mode as well.
+  auto GraphEventD = Q3.submit([&](handler &CGH) {
+    CGH.depends_on(GraphEventB);
+    CGH.parallel_for(range<1>(Size), [=](item<1> Id) { PtrC[Id]++; });
+  });
+
+  assert(Q1.ext_oneapi_get_state() == exp_ext::queue_state::recording);
+  assert(Q2.ext_oneapi_get_state() == exp_ext::queue_state::recording);
+  assert(Q3.ext_oneapi_get_state() == exp_ext::queue_state::recording);
```

Comment:
Slight nit but for this kind of state check I would probably explicitly check all 3 after each call to submit that the state is what it should be. Only checking all of them at the end leaves a bit of space for incorrect behaviour to sneak through.

This might also be better suited as a unit test, do we need to be running kernels on device for this test to function correctly?

Reply:
I think running kernels can help make sure that the feature is working properly. Moving this to a unit test would only check if the queue state changed. But I think that even though the queue state has been updated correctly, the implementation can still be wrong.

I agree that some extra tests as unit tests might be useful though. In particular, I should add negative tests

Reply:
Having both is a good idea and unit tests look good to me!